src/frequenz/sdk/data_ingestion/resampling/component_metrics_resampling_actor.py

```diff
@@ -0,0 +1,240 @@
+
+        self._resampler = ComponentMetricGroupResampler(
+            resampling_period_s=resampling_period_s,
+            max_data_age_in_periods=max_data_age_in_periods,
+            initial_resampling_function=resampling_function,
+        )
+
+        self._input_receivers: Dict[str, Receiver[Sample]] = {}
+        self._output_senders: Dict[str, Sender[Sample]] = {}
+        self._resampling_timer = Timer(interval=self._resampling_period_s)
+
+    async def subscribe(self, request: ComponentMetricRequest) -> None:
+        """Subscribe for data for a specific time series.
+
+        Args:
+            request: subscription request for a specific component metric
+        """
+        channel_name = request.get_channel_name()
+
+        data_source_request = dataclasses.replace(request, **dict(namespace="Source"))
+        data_source_channel_name = data_source_request.get_channel_name()
+        if channel_name not in self._input_receivers:
+            await self._subscription_sender.send(data_source_request)
+            receiver: Receiver[Sample] = self._channel_registry.get_receiver(
+                data_source_channel_name
+            )
+            self._input_receivers[data_source_channel_name] = receiver
+            self._resampler.add_time_series(time_series_id=data_source_channel_name)
+
+        if channel_name not in self._output_senders:
+            sender: Sender[Sample] = self._channel_registry.get_sender(channel_name)
+            # This means that the `sender` will be sending samples to the channel with
+            # name `channel_name` based on samples collected from the channel named
+            # `data_source_channel_name`
+            self._output_senders[data_source_channel_name] = sender
+
+    def is_sample_valid(self, sample: Sample) -> bool:
+        """Check if the provided sample is valid.
+
+        Args:
+            sample: sample to be validated
+
+        Returns:
+            True if the sample is valid, False otherwise
+        """
+        if sample.value is None or math.isnan(sample.value):
+            return False
+        return True
+
+    async def run(self) -> None:
+        """Run the actor.
+
+        Raises:
+            ConnectionError: When the provider of the subscription channel closes the
+                connection
+        """
+        while True:
+            select = Select(
+                resampling_timer=self._resampling_timer,
+                subscription_receiver=self._subscription_receiver,
+                component_data_receiver=MergeNamed(**self._input_receivers),
+            )
+            while await select.ready():
+                if _ := select.resampling_timer:
+                    awaitables = [
+                        self._output_senders[channel_name].send(sample)
+                        for channel_name, sample in self._resampler.resample()
+                    ]
+                    await asyncio.gather(*awaitables)
```

Comment:
the timer actually returns a `datetime` object at which the timer triggered.  We can actually pass it to the `resample` function and it can pass it down the chain.

Reply:
Interesting. I just wonder how likely it is that there is a considerable time difference between the fire of the timer and this processing (for example if there is another "event" processing before this one that took a while) [^1]. I guess it is still more correct to use the timestamp of the fired timer, but then we should be careful also that samples in the *future* are not sent to the resampling function.

I will give it a thought when restructuring stuff to make it more easy to reuse.

[^1]: This also reminds me of a thought I've been having for a while, for which I just created an issue: https://github.com/frequenz-floss/frequenz-channels-python/issues/27

Reply:
https://github.com/frequenz-floss/frequenz-sdk-python/issues/59